be/queue.c

```diff
@@ -0,0 +1,443 @@
+	M0_PRE(!be_queue_is_empty(bq));
+
+	bqi = bqq_tlist_head(&bq->bq_q);
+	m0_buf_memcpy(data, &BE_QUEUE_ITEM2BUF(bq, bqi));
+	*successful = true;
+	bqq_tlist_move(&bq->bq_q_unused, bqi);
+	++bq->bq_dequeued;
+	M0_LEAVE("bq="BEQ_F, BEQ_P(bq));
+}
+
+static void be_queue_op_put(struct m0_be_queue   *bq,
+                            struct m0_be_op      *op,
+                            struct be_queue_item *bqi)
+{
+	struct be_queue_wait_op *bwo;
+
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_PRE(!bqop_tlist_is_empty(&bq->bq_op_put_unused));
+
+	bwo = bqop_tlist_head(&bq->bq_op_put_unused);
+	bwo->bbo_bqi = bqi;
+	bwo->bbo_op  = op;
+	bqop_tlist_move_tail(&bq->bq_op_put, bwo);
+	M0_LEAVE("bq="BEQ_F, BEQ_P(bq));
+}
+
+static void be_queue_op_put_done(struct m0_be_queue *bq)
+{
+	struct be_queue_wait_op *bwo;
+
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_PRE(!bqop_tlist_is_empty(&bq->bq_op_put));
+
+	bwo = bqop_tlist_head(&bq->bq_op_put);
+	m0_be_op_done(bwo->bbo_op);
+	bqop_tlist_move(&bq->bq_op_put_unused, bwo);
+	M0_LEAVE("bq="BEQ_F, BEQ_P(bq));
+}
+
+static bool be_queue_op_put_is_waiting(struct m0_be_queue *bq)
+{
+	return !bqop_tlist_is_empty(&bq->bq_op_put);
+}
+
+static void be_queue_op_get(struct m0_be_queue *bq,
+                            struct m0_be_op    *op,
+                            struct m0_buf      *data,
+                            bool               *successful)
+{
+	struct be_queue_wait_op *bwo;
+
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_PRE(!bqop_tlist_is_empty(&bq->bq_op_get_unused));
+
+	bwo = bqop_tlist_head(&bq->bq_op_get_unused);
+	bwo->bbo_data       = *data;
+	bwo->bbo_successful = successful;
+	bwo->bbo_op         = op;
+	bqop_tlist_move_tail(&bq->bq_op_get, bwo);
+	M0_LEAVE("bq="BEQ_F, BEQ_P(bq));
+}
+
+static void be_queue_op_get_done(struct m0_be_queue *bq, bool success)
+{
+	struct be_queue_wait_op *bwo;
+
+	M0_ENTRY("bq="BEQ_F" success=%d", BEQ_P(bq), !!success);
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_PRE(!bqop_tlist_is_empty(&bq->bq_op_get));
+
+	bwo = bqop_tlist_head(&bq->bq_op_get);
+	if (success) {
+		be_queue_q_get(bq, &bwo->bbo_data, bwo->bbo_successful);
+	} else {
+		*bwo->bbo_successful = false;
+	}
+	m0_be_op_done(bwo->bbo_op);
+	bqop_tlist_move(&bq->bq_op_get_unused, bwo);
+	M0_LEAVE("bq="BEQ_F, BEQ_P(bq));
+}
+
+static bool be_queue_op_get_is_waiting(struct m0_be_queue *bq)
+{
+	return !bqop_tlist_is_empty(&bq->bq_op_get);
+}
+
+M0_INTERNAL void m0_be_queue_put(struct m0_be_queue  *bq,
+                                 struct m0_be_op     *op,
+                                 const struct m0_buf *data)
+{
+	struct be_queue_item *bqi;
+	bool                  was_full;
+
+	M0_ENTRY("bq="BEQ_F, BEQ_P(bq));
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+	M0_PRE(be_queue_invariant(bq));
+	M0_PRE(!bq->bq_the_end);
+	M0_PRE(data->b_nob == bq->bq_cfg.bqc_item_length);
+
+	m0_be_op_active(op);
+	was_full = be_queue_is_full(bq);
+	bqi = be_queue_q_put(bq, data);
```

Comment:
line 347 has queue full check . If queue is full then Line 348 still puts element in queue, is this behavior correct ?

Reply:
It is. Added the explanation to the top of `be/queue.h`.